scalyr_agent/builtin_monitors/kubernetes_monitor.py

```diff
@@ -412,9 +430,428 @@ def _ignore_old_dead_container( container, created_before=None ):
+        @param temporary_error:  The exception generated while warming the pod that indicated a temporary error.
+            Temporary errors results in the pod being blacklisted after __max_failure_count.
+        @param unknown_error:  The exception generated while warming the pod that could not be categorized.
+            It is treated as a temporary error.
+        @param traceback_report:  The traceback of the exception if one of the error types
+
+        @type container_id: str
+        @type success: bool or None
+        @type permanent_error: Exception or None
+        @type temporary_error: Exception or None
+        @type unknown_error: Exception or None
+        @type traceback_report: str or None
+        """
+        current_time = self._get_current_time()
+        result_type = 'not_set'
+        exception_to_report = None
+        self.__lock.acquire()
+        try:
+            if container_id in self.__active_pods:
+                entry = self.__active_pods[container_id]
+                if success:
+                    entry.is_warm = True
+                    entry.failure_count = 0
+                    entry.blacklisted_until = None
+                    entry.blacklist_reason = None
+                    result_type = 'success'
+                else:
+                    if permanent_error:
+                        result_type = 'perm_error'
+                        exception_to_report = permanent_error
+                    elif temporary_error:
+                        result_type = 'temp_error'
+                        exception_to_report = temporary_error
+                    elif unknown_error:
+                        result_type = 'unknown_error'
+                        exception_to_report = unknown_error
+                    else:
+                        result_type = 'unhandled_error'
+
+                    entry.failure_count += 1
+                    if permanent_error or entry.failure_count >= self.__max_failure_count:
+                        entry.blacklisted_until = current_time + self.__blacklist_time_secs
+                        entry.blacklist_reason = result_type
+                self.__update_containers_to_warm()
+
+        finally:
+            self.__warming_attempts['total'] = self.__warming_attempts.get('total', 0) + 1
+            self.__warming_attempts[result_type] = self.__warming_attempts.get(result_type, 0) + 1
+
+            self.__lock.release()
+            if exception_to_report is not None and self.__logger is not None:
+                self.__logger.error('An error of type %s was seen when warming container %s.  Exception was %s: %s',
+                                    result_type, container_id, str(exception_to_report), str(traceback_report),
+                                    limit_once_per_x_secs=300, limit_key='warmer-record-result-%s' % result_type)
+
+    @staticmethod
+    def _get_current_time():
+        """Returns the current time.
+
+        This is used for testing.
```

Comment:
nit: more precisely -- "This method exists so it can be overridden during testing."